frame/collator-selection/src/lib.rs

```diff
@@ -0,0 +1,518 @@
+        }
+
+        #[pallet::weight(T::WeightInfo::set_desired_candidates())]
+        pub fn set_desired_candidates(
+            origin: OriginFor<T>,
+            max: u32,
+        ) -> DispatchResultWithPostInfo {
+            T::UpdateOrigin::ensure_origin(origin)?;
+            // we trust origin calls, this is just a for more accurate benchmarking
+            if max > T::MaxCandidates::get() {
+                log::warn!("max > T::MaxCandidates; you might need to run benchmarks again");
+            }
+            <DesiredCandidates<T>>::put(&max);
+            Self::deposit_event(Event::NewDesiredCandidates(max));
+            Ok(().into())
+        }
+
+        #[pallet::weight(T::WeightInfo::set_candidacy_bond())]
+        pub fn set_candidacy_bond(
+            origin: OriginFor<T>,
+            bond: BalanceOf<T>,
+        ) -> DispatchResultWithPostInfo {
+            T::UpdateOrigin::ensure_origin(origin)?;
+            <CandidacyBond<T>>::put(&bond);
+            Self::deposit_event(Event::NewCandidacyBond(bond));
+            Ok(().into())
+        }
+
+        #[pallet::weight(T::WeightInfo::register_as_candidate(T::MaxCandidates::get()))]
+        pub fn register_as_candidate(origin: OriginFor<T>) -> DispatchResultWithPostInfo {
+            let who = ensure_signed(origin)?;
+
+            // ensure we are below limit.
+            let length = <Candidates<T>>::decode_len().unwrap_or_default();
+            ensure!(
+                (length as u32) < Self::desired_candidates(),
+                Error::<T>::TooManyCandidates
+            );
+            ensure!(
+                !Self::invulnerables().contains(&who),
+                Error::<T>::AlreadyInvulnerable
+            );
+
+            let validator_key = T::ValidatorIdOf::convert(who.clone())
+                .ok_or(Error::<T>::NoAssociatedValidatorId)?;
+            ensure!(
+                T::ValidatorRegistration::is_registered(&validator_key),
+                Error::<T>::ValidatorNotRegistered
+            );
+
+            let deposit = Self::candidacy_bond();
+
+            let free_balance = T::Currency::free_balance(&who);
+            ensure!(free_balance > deposit, Error::<T>::TooLowFreeBalance);
+
+            // First authored block is current block plus kick threshold to handle session delay
+            let incoming = CandidateInfo {
+                who: who.clone(),
+                deposit,
+            };
+
+            let current_count =
+                <Candidates<T>>::try_mutate(|candidates| -> Result<usize, DispatchError> {
+                    if candidates.into_iter().any(|candidate| candidate.who == who) {
+                        Err(Error::<T>::AlreadyCandidate)?
+                    } else {
+                        T::Currency::set_lock(
+                            COLLATOR_STAKING_ID,
+                            &who,
+                            deposit,
+                            WithdrawReasons::all(),
```

Comment:
So this means that it doesn't matter how much ASTR the candidate has?
Interesting.

Reply:
Technically one address can't be collator twice, in this case this is correct. We just need to have minimum to join.

Reply:
Sorry, I missed the part at L:357 where you do the check.